boot/freeldr/freeldr/arch/uefi/uefidisk.c

```diff
@@ -0,0 +1,591 @@
+            continue;
+        }
+        if (bio->Media->LogicalPartition == FALSE)
+        {
+            TRACE("HardDisk\n");
+            PcBiosDiskCount++;
+            InternalUefiDisk[BlockDeviceIndex].ArcDriveNumber = BlockDeviceIndex;
+            InternalUefiDisk[BlockDeviceIndex].UefiRootNumber = i;
+            GetHarddiskInformation(BlockDeviceIndex + FIRST_BIOS_DISK);
+            BlockDeviceIndex++;
+        }
+        else if (handles[i] == PublicBootHandle)
+        {
+            ULONG increment = 0;
+            ULONG i;
+
+            /* Grab the offset into the array of handles and decrement per volume (valid partition)*/
+            for (increment = OffsetToBoot; increment > 0; increment--)
+            {
+                GlobalSystemTable->BootServices->HandleProtocol(handles[increment], &bioGuid, (void**)&bio);
+                if (bio->Media->LogicalPartition == FALSE)
+                {
+                    TRACE("Found root at increment %u", increment);
+                    UefiBootRootIdentifier = increment;
+
+                    for(i = 0; i <= PcBiosDiskCount; ++i)
+                    {
+                        /* Now only of the root drive number is equal to this drive we found above */
+                        if (InternalUefiDisk[i].UefiRootNumber == UefiBootRootIdentifier)
+                        {
+                            InternalUefiDisk[i].IsThisTheBootDrive == TRUE;
+                            PublicBootArcDisk = i;
+                            TRACE("Found Boot drive\n");
+                        }
+                    }
+
+                    break;
+                }
+            }
+        }
+    }
+}
+
+static
+BOOLEAN
+UefiSetBootpath(VOID)
+{
+   TRACE("UefiSetBootpath: Setting up boot path\n");
+   GlobalSystemTable->BootServices->HandleProtocol(handles[UefiBootRootIdentifier], &bioGuid, (void**)&bio);
+   FrldrBootDrive = (FIRST_BIOS_DISK + PublicBootArcDisk);
+   if (bio->Media->RemovableMedia == TRUE && bio->Media->BlockSize == 2048)
+   {
+        /* Boot Partition 0xFF is the magic value that indicates booting from CD-ROM (see isoboot.S) */
+        FrldrBootPartition == 0xFF;
+        RtlStringCbPrintfA(FrLdrBootPath, sizeof(FrLdrBootPath),
+                           "multi(0)disk(0)cdrom(%u)", PublicBootArcDisk);
+   }
+   else
+   {
+        ULONG BootPartition;
+        PARTITION_TABLE_ENTRY PartitionEntry;
+
+        /* This is a hard disk */
+        if (!UefiGetBootPartitionEntry(FrldrBootDrive, &PartitionEntry, &BootPartition))
+        {
+            ERR("Failed to get boot partition entry\n");
+            return FALSE;
+        }
+
+        RtlStringCbPrintfA(FrLdrBootPath, sizeof(FrLdrBootPath),
+                           "multi(0)disk(0)rdisk(%u)partition(%lu)",
+                           PublicBootArcDisk, BootPartition);
+    }
+
+    return TRUE;
+}
+
+BOOLEAN
+UefiInitializeBootDevices(VOID)
+{
+    ULONG i = 0;
+
+    DiskReadBufferSize = EFI_PAGE_SIZE;
```

Comment:
Indentation broken (tabs instead of spaces)

Reply:
Where is it using tabs?